### PR TITLE
Avoid allocs improving ~15% performance and using ~97% less memory

### PR DESCRIPTION
**Pull request by jsign**
This PR contains a set of optimizations that improves the performance of the library significantly.

Using the reference benchmarks in the repo, here's a comparison in performance between `master` and this PR:
```
name                           old time/op    new time/op    delta
pkg:github.com/crate-crypto/go-ipa/bandersnatch goos:linux goarch:amd64
MultiExpG1/32_points-16           318µs ± 6%     355µs ±11%   +11.81%  (p=0.000 n=10+8)
MultiExpG1/64_points-16           354µs ± 5%     287µs ± 4%   -18.69%  (p=0.000 n=10+7)
MultiExpG1/128_points-16          450µs ± 5%     381µs ± 6%   -15.31%  (p=0.000 n=10+8)
MultiExpG1/256_points-16          649µs ± 4%     556µs ± 5%   -14.31%  (p=0.000 n=10+8)
MultiExpG1/512_points-16         1.04ms ± 4%    0.87ms ± 2%   -16.34%  (p=0.000 n=9+8)
MultiExpG1/1024_points-16        1.63ms ± 7%    1.30ms ± 3%   -20.39%  (p=0.000 n=10+8)
MultiExpG1/2048_points-16        2.85ms ± 3%    2.40ms ± 1%   -15.69%  (p=0.000 n=10+7)
MultiExpG1/4096_points-16        4.98ms ± 2%    4.17ms ± 1%   -16.37%  (p=0.000 n=10+8)
MultiExpG1/8192_points-16        9.49ms ± 0%    8.06ms ± 0%   -15.09%  (p=0.000 n=10+7)
MultiExpG1/16384_points-16       18.7ms ± 1%    15.7ms ± 1%   -15.83%  (p=0.000 n=10+8)
MultiExpG1/32768_points-16       36.0ms ± 2%    30.8ms ± 2%   -14.45%  (p=0.000 n=10+8)
MultiExpG1/65536_points-16       64.1ms ± 3%    54.3ms ± 2%   -15.25%  (p=0.000 n=10+7)
MultiExpG1/131072_points-16       109ms ± 3%      93ms ± 2%   -14.43%  (p=0.000 n=10+8)
MultiExpG1/262144_points-16       192ms ± 2%     164ms ± 7%   -14.70%  (p=0.000 n=10+8)
MultiExpG1/524288_points-16       381ms ± 2%     321ms ± 2%   -15.75%  (p=0.000 n=10+7)
MultiExpG1/1048576_points-16      722ms ± 3%     605ms ± 2%   -16.21%  (p=0.000 n=10+8)
MultiExpG1/2097152_points-16      1.41s ± 2%     1.19s ± 1%   -15.71%  (p=0.000 n=10+8)
MultiExpG1/4194304_points-16      2.82s ± 1%     2.38s ± 1%   -15.57%  (p=0.000 n=10+8)
MultiExpG1/8388608_points-16      5.69s ± 0%     5.04s ± 1%   -11.41%  (p=0.000 n=9+7)
MultiExpG1/16777216_points-16     10.6s ± 1%      9.2s ± 0%   -12.93%  (p=0.000 n=10+8)
MultiExpG1Reference-16            741ms ± 4%     608ms ± 0%   -18.04%  (p=0.000 n=9+6)
ManyMultiExpG1Reference-16        2.18s ± 2%     1.81s ± 1%   -17.23%  (p=0.000 n=10+8)

name                           old alloc/op   new alloc/op   delta
pkg:github.com/crate-crypto/go-ipa/bandersnatch goos:linux goarch:amd64
MultiExpG1/32_points-16           231kB ± 0%      18kB ± 0%   -92.38%  (p=0.000 n=9+8)
MultiExpG1/64_points-16           347kB ± 0%      21kB ± 0%   -94.06%  (p=0.000 n=8+8)
MultiExpG1/128_points-16          551kB ± 0%      23kB ± 0%   -95.89%  (p=0.000 n=9+8)
MultiExpG1/256_points-16          905kB ± 0%      24kB ± 0%   -97.37%  (p=0.000 n=10+8)
MultiExpG1/512_points-16         1.54MB ± 0%    0.03MB ± 0%   -98.06%  (p=0.000 n=10+8)
MultiExpG1/1024_points-16        2.67MB ± 0%    0.04MB ± 0%   -98.33%  (p=0.000 n=9+8)
MultiExpG1/2048_points-16        4.77MB ± 0%    0.08MB ± 0%   -98.40%  (p=0.000 n=10+8)
MultiExpG1/4096_points-16        8.59MB ± 0%    0.14MB ± 0%   -98.35%  (p=0.000 n=8+8)
MultiExpG1/8192_points-16        15.5MB ± 0%     0.3MB ± 0%   -98.25%  (p=0.000 n=10+8)
MultiExpG1/16384_points-16       28.1MB ± 0%     0.5MB ± 0%   -98.10%  (p=0.000 n=9+8)
MultiExpG1/32768_points-16       52.1MB ± 0%     1.1MB ± 0%   -97.97%  (p=0.000 n=10+8)
MultiExpG1/65536_points-16       96.0MB ± 0%     2.1MB ± 0%   -97.81%  (p=0.000 n=10+8)
MultiExpG1/131072_points-16       180MB ± 0%       4MB ± 0%   -97.67%  (p=0.000 n=10+8)
MultiExpG1/262144_points-16       329MB ± 0%       8MB ± 0%   -97.45%  (p=0.000 n=10+7)
MultiExpG1/524288_points-16       621MB ± 0%      17MB ± 0%   -97.30%  (p=0.000 n=10+8)
MultiExpG1/1048576_points-16     1.17GB ± 0%    0.03GB ± 0%   -97.14%  (p=0.000 n=10+8)
MultiExpG1/2097152_points-16     2.28GB ± 0%    0.07GB ± 0%   -97.06%  (p=0.000 n=10+8)
MultiExpG1/4194304_points-16     4.56GB ± 0%    0.13GB ± 0%   -97.06%  (p=0.000 n=8+8)
MultiExpG1/8388608_points-16     10.1GB ± 0%     1.5GB ± 0%   -85.35%  (p=0.000 n=9+8)
MultiExpG1/16777216_points-16    17.3GB ± 0%     1.7GB ± 0%   -89.93%  (p=0.000 n=10+8)
MultiExpG1Reference-16           1.17GB ± 0%    0.03GB ± 0%   -97.14%  (p=0.000 n=10+7)
ManyMultiExpG1Reference-16       3.52GB ± 0%    0.10GB ± 0%   -97.14%  (p=0.000 n=8+8)

name                           old allocs/op  new allocs/op  delta
pkg:github.com/crate-crypto/go-ipa/bandersnatch goos:linux goarch:amd64
MultiExpG1/32_points-16           3.39k ± 0%     0.09k ± 0%   -97.38%  (p=0.000 n=10+8)
MultiExpG1/64_points-16           5.22k ± 0%     0.13k ± 0%   -97.51%  (p=0.000 n=10+8)
MultiExpG1/128_points-16          8.38k ± 0%     0.13k ± 0%   -98.45%  (p=0.000 n=10+8)
MultiExpG1/256_points-16          13.9k ± 0%      0.1k ± 0%   -99.19%  (p=0.000 n=10+8)
MultiExpG1/512_points-16          23.7k ± 0%      0.1k ± 0%   -99.58%  (p=0.000 n=10+8)
MultiExpG1/1024_points-16         41.2k ± 0%      0.1k ± 0%   -99.78%  (p=0.000 n=10+8)
MultiExpG1/2048_points-16         73.4k ± 0%      0.1k ± 0%   -99.89%  (p=0.000 n=10+8)
MultiExpG1/4096_points-16          132k ± 0%        0k ± 0%   -99.94%  (p=0.000 n=10+8)
MultiExpG1/8192_points-16          238k ± 0%        0k ± 0%   -99.97%  (p=0.000 n=10+8)
MultiExpG1/16384_points-16         431k ± 0%        0k ± 0%   -99.98%  (p=0.000 n=10+8)
MultiExpG1/32768_points-16         798k ± 0%        0k ± 0%   -99.99%  (p=0.000 n=10+8)
MultiExpG1/65536_points-16        1.47M ± 0%     0.00M ± 0%  -100.00%  (p=0.000 n=10+8)
MultiExpG1/131072_points-16       2.75M ± 0%     0.00M ± 0%  -100.00%  (p=0.000 n=10+8)
MultiExpG1/262144_points-16       5.01M ± 0%     0.00M ± 0%  -100.00%  (p=0.000 n=10+8)
MultiExpG1/524288_points-16       9.44M ± 0%     0.00M ± 0%  -100.00%  (p=0.000 n=10+8)
MultiExpG1/1048576_points-16      17.8M ± 0%      0.0M ± 0%  -100.00%  (p=0.000 n=10+8)
MultiExpG1/2097152_points-16      34.6M ± 0%      0.0M ± 1%  -100.00%  (p=0.000 n=10+8)
MultiExpG1/4194304_points-16      69.2M ± 0%      0.0M ± 1%  -100.00%  (p=0.000 n=10+8)
MultiExpG1/8388608_points-16       134M ± 0%        0M ± 1%  -100.00%  (p=0.000 n=10+8)
MultiExpG1/16777216_points-16      243M ± 0%        0M ± 3%  -100.00%  (p=0.000 n=10+7)
MultiExpG1Reference-16            17.8M ± 0%      0.0M ± 0%  -100.00%  (p=0.000 n=10+8)
ManyMultiExpG1Reference-16        53.5M ± 0%      0.0M ± 2%  -100.00%  (p=0.000 n=9+8)
```
A TL;DR:
- Around 15% reduction in computation time in most sub-tests, with 32-points being a bit of an outlier.
- The performance gain comes from avoiding allocations, which clearly is shown by reducing by more than 95% and some functions completely stopping doing memory allocations.

The original focus to guide the optimizations was making `go-verkle` faster for generating and verifying proofs.

To understand the impact of these optimizations in `go-verkle` see [this PR](https://github.com/gballet/go-verkle/pull/213).

I'm not entirely sure if the repo is open for PRs from external contributors; if this PR is looking good to consider for merging I can add some comments on the changes.  

Let me know!

### CODE REVIEW

**Comment by kevaundray** [comment]
These are amazing results! Thank you very much,  I'll check out the PR now :)

File: bandersnatch/multiexp.go [hunk 1]
```diff
@@ -448,37 +478,48 @@ func (p *PointProj) msmC4(points []PointAffine, scalars []fr.Element, splitFirst
 	// critical for performance
 
 	// each go routine sends its result in chChunks[i] channel
-	var chChunks [nbChunks]chan PointProj
-	for i := 0; i < len(chChunks); i++ {
-		chChunks[i] = make(chan PointProj, 1)
-	}
-
-	processChunk := func(j int, points []PointAffine, scalars []fr.Element, chChunk chan PointProj) {
+	//var chChunks [nbChunks]chan PointProj
```
Comment (kevaundray) [inline bandersnatch/multiexp.go:481]: unused code chunk here

File: bandersnatch/point.go [hunk 2]
```diff
@@ -104,17 +104,16 @@ func (p *PointAffine) SetBytes(buf []byte) (int, error) {
 
 // Reads an uncompressed affine point
 // Point is not guaranteed to be in the prime subgroup 
-func ReadUncompressedPoint(r io.Reader) *PointAffine {
-	var x = make([]byte, 32)
-	var y = make([]byte, 32)
-	n, err := r.Read(x)
+func ReadUncompressedPoint(r io.Reader) PointAffine {
+	var xy = make([]byte, 64)
```
Comment (kevaundray) [inline bandersnatch/point.go:108]: This affects readability a bit, but I'm guessing that because we Read alot, making one allocation showed significant performance gains?

File: bandersnatch/precomp_multiexp.go [hunk 3]
```diff
@@ -75,52 +73,37 @@ func DeserializePrecomputedLagrange(reader io.Reader) (*PrecomputeLagrange, erro
...
 	return &pcl, nil
 }
 
 func (p *PrecomputeLagrange) Commit(evaluations []fr.Element) *PointAffine {
+	var result PointProj
+	result.Identity()
 
-	nbTasks := runtime.NumCPU()
-	chPartialResults := make(chan PointProj, nbTasks)
-	parallel.Execute(len(evaluations), func(start, end int) {
```
Comment (kevaundray) [inline bandersnatch/precomp_multiexp.go:89]: Interesting, why did you remove the parallel code here? Did you see a performance degradation?

**Comment by kevaundray** [comment]
Oh RE your last sentence, yes please add comments on the changes!

**Comment by kevaundray** [comment]
I should also note that #19 might regress some of the changes made here. Instead of PointAffine, we use a wrapper around PointProj with special methods, so MultiExp etc now also use that wrapper

File: bandersnatch/multiexp.go (hunk 1 above)
Comment (jsign) [reply bandersnatch/multiexp.go]: I left these comments here since this was still a draft PR to get a temperature check of the proposed changes. 
I'll clean this for sure :+1:

File: bandersnatch/multiexp.go [hunk 4]
```diff
@@ -367,12 +369,40 @@ func msmReduceChunkPointAffine(p *PointProj, c int, chChunks []chan PointProj) *
...
 	buckets []PointProj,
 	c uint64,
 	points []PointAffine,
 	scalars []fr.Element) {
+	var total PointProj
+	msmProcessChunkPointAffineDMA(chunk, &total, buckets, c, points, scalars)
+	chRes <- total
+}
+
+func msmProcessChunkPointAffineDMA(chunk uint64,
```
Comment (jsign) [inline bandersnatch/multiexp.go:400]: The TL;DR in the multiexp is that a channel-based solution to the "map->reduce"  created many allocations regarding creating so many channels.

I noticed that the way the work was structured, it could be directly mapped to a slice index avoiding all that extra work.

The "reduce" code is backward compatible since there's another code that expects to receive the result in a channel.  (Code above this comment).

File: bandersnatch/point.go (hunk 2 above)
Comment (jsign) [reply bandersnatch/point.go:108]: Correct, this was at the hot path of `go-verkle` benchmarks. 

Also, something relevant to say about this is that both `x` and `y` have the same "lifetime", so bundling them in the same slice can't have unwanted consequences regarding GC. 

Note also that I avoid the pointer return since that confuses the compiler in escaping analysis.

File: bandersnatch/point.go [hunk 5]
```diff
@@ -216,19 +213,17 @@ func NewPointAffine(x, y fp.Element) PointAffine {
 // IsOnCurve checks if a point is on the twisted Edwards curve
 func (p *PointAffine) IsOnCurve() bool {
 
-	ecurve := GetEdwardsCurve()
-
```
Comment (jsign) [inline bandersnatch/point.go:220]: The problem here is that `GetEdwardsCurve` forces the construction to allocate a `big.Int`, which escapes to the heap.
But, most of the callers aren't really interested in the `order` field, but in immutable fields of the curve.

This was the top 1 source of allocs.

File: bandersnatch/precomp_multiexp.go (hunk 3 above)
Comment (jsign) [reply bandersnatch/precomp_multiexp.go:89]: The `parallel.Execute` has some unfortunate heap allocations due to the nature of spinning up goroutines and work partitioning. First, I tried to see if that could be avoided but wasn't really possible.

But then I simply challenged the assumption that this work really needed parallel work. So took some benchmarks about how heavy was each operation and was not big enough to justify the overhead of partitioning and goroutines.

The unparalleled execution turned out to be faster, and avoid the CPU overhead plus indirect allocations of the parallelization.

**Comment by jsign** [comment]
> These are amazing results! Thank you very much, I'll check out the PR now :)

You're welcome!

> I should also note that https://github.com/crate-crypto/go-ipa/pull/19 might regress some of the changes made here. Instead of PointAffine, we use a wrapper around PointProj with special methods, so MultiExp etc now also use that wrapper

Sounds good. Maybe after that gets merged I can take some other pass with optimization-eyes to see if there's something extra that we can gain.

File: bandersnatch/precomp_multiexp.go (hunk 3 above)
Comment (kevaundray) [reply bandersnatch/precomp_multiexp.go:89]: Yeah it makes sense that for small multi exponentations (256), the parallel code might incur a non-negligble overhead

**Comment by kevaundray** [comment]
Alright, I'll merge this first and then resolve conflicts on that branch, aiming to preserve the optimisations you have done

**Review by kevaundray — APPROVED** [review]
